### PR TITLE
Feature/unit test

### PR DESCRIPTION
**Pull request by SteMat89**

### CODE REVIEW

File: Chessington.GameEngine/Pieces/Pawn.cs [hunk 1]
```diff
@@ -14,11 +14,66 @@ public override IEnumerable<Square> GetAvailableMoves(Board board)
             
             if (this.Player == Player.White)
             {
-                availableMoves.Add(new Square(board.FindPiece(this).Row - 1, board.FindPiece(this).Col));
+                if (board.FindPiece(this).Row == 6
```
Comment (AndreiHoszu) [inline Chessington.GameEngine/Pieces/Pawn.cs]: this part could have been integrated in the if below

File: Chessington.GameEngine/Pieces/Pawn.cs [hunk 2]
```diff
@@ -14,11 +14,66 @@ public override IEnumerable<Square> GetAvailableMoves(Board board)
...
+                        board.GetPiece(new Square(board.FindPiece(this).Row - 1, board.FindPiece(this).Col - 1)) != null 
+                        &&  board.GetPiece(new Square(board.FindPiece(this).Row - 1, 
+                            board.FindPiece(this).Col - 1)).Player != this.Player)
+                    {
+                        availableMoves.Add(new Square(board.FindPiece(this).Row - 1, board.FindPiece(this).Col - 1));
+                    }
+                }
             }
-            else if (this.Player == Player.Black)
+            else if (this.Player == Player.Black )
```
Comment (AndreiHoszu) [inline Chessington.GameEngine/Pieces/Pawn.cs]: redundant if condition

File: Chessington.GameEngine/Pieces/King.cs [hunk 3]
```diff
@@ -10,7 +10,27 @@ public King(Player player)
...
         {
-            return Enumerable.Empty<Square>();
+            List<Square> availableMoves = new List<Square>();
+            
+            int[] row = new int[] { 1, 1, 1, 0, 0, -1, -1, -1};
+            int[] col = new int[] { -1, 0, 1, -1, 1, -1, 0, 1};
+            
+            for (int i = 0; i < 8; i++)
+            {
+                if(board.FindPiece(this).Row + row[i] >= 0
```
Comment (AndreiHoszu) [inline Chessington.GameEngine/Pieces/King.cs]: board.FindPiece(this) could be replaced with a variable above to avoid multiple function calls

Comment (AndreiHoszu) [reply Chessington.GameEngine/Pieces/King.cs]: this is also true for the other classes

Comment (SteMat89) [reply Chessington.GameEngine/Pieces/King.cs]: Yessir that is true. Would've saved me some time and space

File: Chessington.GameEngine/Pieces/Pawn.cs [hunk 4]
```diff
@@ -14,11 +14,66 @@ public override IEnumerable<Square> GetAvailableMoves(Board board)
...
+                {
+                    if (board.GetPiece(new Square(board.FindPiece(this).Row - 1, board.FindPiece(this).Col)) == null)
+                    {
+                        availableMoves.Add(new Square(board.FindPiece(this).Row - 1, board.FindPiece(this).Col));
+                    }
+ 
+                    if (board.FindPiece(this).Col + 1 < 7 && 
+                        board.GetPiece(new Square(board.FindPiece(this).Row - 1, board.FindPiece(this).Col + 1)) != null 
+                        &&  board.GetPiece(new Square(board.FindPiece(this).Row - 1, 
+                            board.FindPiece(this).Col + 1)).Player != this.Player)
```
Comment (AndreiHoszu) [inline Chessington.GameEngine/Pieces/Pawn.cs:46]: this sort of indentation is kind of hard to follow :D, i suggest using the same type of rule: for example here you have the first condition followed by the logical operator, then on one single line you got only a condition, then on the next two lines you got the logical operator first then a condition split because of function arguments, so it is kind of hard to follow

Comment (AndreiHoszu) [reply Chessington.GameEngine/Pieces/Pawn.cs:46]: so something like:

                    if (board.FindPiece(this).Col + 1 < 7
                        &&  board.GetPiece(new Square(board.FindPiece(this).Row - 1, 
                            board.FindPiece(this).Col + 1)) != null 
                        &&  board.GetPiece(new Square(board.FindPiece(this).Row - 1, 
                            board.FindPiece(this).Col + 1)).Player != this.Player)

would have been clearer, or at least would have followed some sort of rule

Comment (SteMat89) [reply Chessington.GameEngine/Pieces/Pawn.cs:46]: That does make a lot of sense. Tried being consistent, but I guess I've missed some places. Also the lines could be shorter by using a variable for board.FindPiece(this) as you've mentioned earlier! Thank you

File: Chessington.GameEngine/Pieces/Rook.cs [hunk 5]
```diff
@@ -10,7 +10,77 @@ public Rook(Player player)
...
+                else if (board.GetPiece(new Square(i, board.FindPiece(this).Col)).Player != this.Player)
+                {
+                    availableMoves.Add(new Square(i, board.FindPiece(this).Col));
+                    break;
+                }
+                else
+                {
+                    break;
+                }
+            }
```
Comment (AndreiVener) [inline Chessington.GameEngine/Pieces/Rook.cs:81]: In order to avoid repeating code you could make a function that checks the available squares in a specified direction eg. List<Squares> GoInDirection(Square startingPosition, int colDir, int rowDir). For the Rook to check all the right squares the colDir = 1 and rowDir = 0.

File: Chessington.GameEngine/Pieces/Knight.cs [hunk 6]
```diff
@@ -10,7 +10,27 @@ public Knight(Player player)
...
+            int[] row = new int[] { 2, 1, -1, -2, -2, -1, 1, 2};
+            int[] col = new int[] { 1, 2, 2, 1, -1, -2, -2, -1};
+
+            for (int i = 0; i < 8; i++)
+            {
+                if (board.FindPiece(this).Row + row[i] >= 0
+                    && board.FindPiece(this).Row + row[i] < 8
+                    && board.FindPiece(this).Col + col[i] >= 0
+                    && board.FindPiece(this).Col + col[i] < 8 
+                    && (board.GetPiece(new Square(board.FindPiece(this).Row + row[i],
```
Comment (PetSteg) [inline Chessington.GameEngine/Pieces/Knight.cs:24]: You're repeating "board.FindPiece(this).Row + row[i]" quite a few times. I'd suggest using a variable instead

File: Chessington.GameEngine/Pieces/Bishop.cs [hunk 7]
```diff
@@ -10,7 +10,76 @@ public Bishop(Player player)
...
+                {
+                    availableMoves.Add(new Square(i, j));
+                    break;
+                }
+                else
+                {
+                    break;
+                }
+            }
+
```
Comment (AndreiVener) [inline Chessington.GameEngine/Pieces/Bishop.cs]: If you implement a GoInDirection(Square startingPosition, int colDir, int rowDir) function (the same from the Rook ) you could shorten this code again eg. If you want the bishop to go in the direction up-right rowDir = -1 and colDir = 1.

File: Chessington.GameEngine/Pieces/Bishop.cs [hunk 8]
```diff
@@ -10,7 +10,76 @@ public Bishop(Player player)
 
         public override IEnumerable<Square> GetAvailableMoves(Board board)
         {
-            return Enumerable.Empty<Square>();
+            List<Square> availableMoves = new List<Square>();
+
+            for (int i = board.FindPiece(this).Row + 1, j = board.FindPiece(this).Col + 1; i < 8 && j < 8; i++, j++)
```
Comment (BogRad97) [inline Chessington.GameEngine/Pieces/Bishop.cs]: Probably more clear if you rename i to row and j to col. Also, we have a `GameSettings.BoardSize`, so use that instead of magic number

File: Chessington.GameEngine/Pieces/King.cs [hunk 9]
```diff
@@ -10,7 +10,27 @@ public King(Player player)
...
+                   && board.FindPiece(this).Row + row[i] < 8 
+                   && board.FindPiece(this).Col + col[i] >= 0 
+                   && board.FindPiece(this).Col + col[i] < 8
```
Comment (BogRad97) [inline Chessington.GameEngine/Pieces/King.cs]: I would make a board.IsSquareInBounds(Square) method, to move this logic (and hopefully make it prettier with several if cases). I imagine this could come in handy in a lot of places

File: Chessington.GameEngine/Pieces/King.cs [hunk 10]
```diff
@@ -10,7 +10,27 @@ public King(Player player)
...
+            int[] row = new int[] { 1, 1, 1, 0, 0, -1, -1, -1};
+            int[] col = new int[] { -1, 0, 1, -1, 1, -1, 0, 1};
+            
+            for (int i = 0; i < 8; i++)
+            {
+                if(board.FindPiece(this).Row + row[i] >= 0
+                   && board.FindPiece(this).Row + row[i] < 8 
+                   && board.FindPiece(this).Col + col[i] >= 0 
+                   && board.FindPiece(this).Col + col[i] < 8 
+                   && (board.GetPiece(new Square(board.FindPiece(this).Row + row[i],
```
Comment (BogRad97) [inline Chessington.GameEngine/Pieces/King.cs]: I would also make helper board.IsSquareEmpty(Square) and board.IsSquareOccupiedByEnemy(Square), with names TBD, because it seems you reuse this logic in several places

Comment (BogRad97) [reply Chessington.GameEngine/Pieces/King.cs]: Or other helper function which is combination of the 2, if you think that's better in your case

File: Chessington.GameEngine/Pieces/Pawn.cs [hunk 11]
```diff
@@ -14,11 +14,66 @@ public override IEnumerable<Square> GetAvailableMoves(Board board)
             
             if (this.Player == Player.White)
```
Comment (BogRad97) [inline Chessington.GameEngine/Pieces/Pawn.cs:25]: This is a very long method that's difficult to follow. Consider splitting it into 2 or more private methods

File: Chessington.GameEngine/Pieces/Queen.cs [hunk 12]
```diff
@@ -9,8 +9,145 @@ public Queen(Player player)
...
 
         public override IEnumerable<Square> GetAvailableMoves(Board board)
-        {
-            return Enumerable.Empty<Square>();
+        { 
+            List<Square> availableMoves = new List<Square>();
+            
+            for (int i = board.FindPiece(this).Row + 1, j = board.FindPiece(this).Col + 1; i < 8 && j < 8; i++, j++)
+            {
+                if (board.GetPiece(new Square(i, j)) == null)
```
Comment (BogRad97) [inline Chessington.GameEngine/Pieces/Queen.cs:17]: You can extract the code inside the for loops into a function, because it looks the same, right?

Comment (BogRad97) [reply Chessington.GameEngine/Pieces/Queen.cs:17]: You could even extract it into a Piece class, since it's used by multiple pieces
